module/wrappers.py

```diff
@@ -13,20 +13,26 @@
 
 def IDWrapper(freer):
     """
-    Classes create with IDWrapper return an ID and then free it upon exit of the
-    context.
+    Classes create with IDWrapper return an ID and then free it upon exit of
+    the context or when garbage collected, whichever comes first.
     """
     class Wrapper:
         def __init__(self, conn):
-            self.conn = conn
-            self.id = None
+            super().__init__()
+            self.id = conn.generate_id()
+            self._free = lambda self=self: getattr(conn.core, freer)(self.id)
+            self.freed = False
 
         def __enter__(self):
-            self.id = self.conn.generate_id()
             return self.id
 
         def __exit__(self, exception_type, exception_value, traceback):
-            getattr(self.conn.core, freer)(self.id)
+            self._free()
+            self.freed = True
+
+        def __del__(self):
```

Comment:
Unfortunately, I think this will result in leaks. X11 resources are not associated with a specific xcb connection, and will not be automatically freed when the connection that created them closes. Python does not guarantee calling `__del__` when the interpreter is exiting, so we may leak an allocated resource by using `__del__` like this.

Reply:
I figured out a workaround - we can use weakref.finalize to add a last ditch effort finalizer that is guaranteed to be called on python exit if for some reason the object is not gced.  This is the same thing that tempfile.TemporaryDirectory does to ensure the directory is cleaned up no matter what.

Reply:
I see,

> A finalizer will never invoke its callback during the later part of the interpreter shutdown when module globals are liable to have been replaced by None.

from https://docs.python.org/3/library/weakref.html

so can't we still leak things even with this approach?

Reply:
as far as I understand it, that's saying that the callbacks are guaranteed to be called before that point... but I will confirm that.

Reply:
OH, sorry, I said weakref.finalize but it also requires atexit.register

Reply:
https://stackoverflow.com/questions/6104535/i-dont-understand-this-python-del-behaviour